### ISSUE TITLE
Extensions for starting with custom MinecraftInterface

### ISSUE THREAD

**Issue by fraenkelc** [toolbox4minecraft/amidst#369]
I've hacked together a working integration of Amidst into a Forge client ( see https://github.com/fraenkelc/amidst-forge ). The interfaces made it straight forward to call the real BiomeProvider / GenLayer code and use that. What gave me some issues was supplying that custom `MinecraftInterface` instance to the amidst client. What I've come up with can be seen in https://github.com/fraenkelc/amidst/commit/a5dca69336771ef4805ff745af4d908fbc062ec1 .
It's neither pretty, nor complete; I'd like to have the option of disabling the profile switcher and load from save game menu items for example.

Is there any interest in creating / integrating an interface for that usecase? I thought I'd raise the question here before doing more work on the amidst codebase.

**Comment by stefandollase** [comment]
Thanks for your effort, I like the idea. However, I am a bit afraid to put Amidst into a new execution context, since I am not familiar with forge and I am sure there will be bugs which are related to putting Amidst into the new execution context. Any thoughts about this?

**Comment by fraenkelc** [comment]
Embedding amidst and running it in the same process just was the easiest way to get the integration running and it required the least modifications to amidst. You're right, there's various things forge does that might cause issues, such as paching minecraft and java classes, using the minecraft logger and also registering its own uncaught exception handler.

A different way of integrating amidst with a modded client could be achieved using a remote connection with a lightweight binary protocol. If that's too slow memory mapped files might be an option. That way Amidst could run in its own JVM. This approach would require some extensions to the MinecraftInterface to handle updates to the biome list.

**Comment by fraenkelc** [comment]
Time to give this a poke a year later. There is some interest in running amidst against modded instances - the most recent version of my fork had 1800 downloads in the last 4 months.

 Is there any interest in adding a remoting interface and extending amidst to accept a biome list and biome data from a running modded minecraft instance? If so, how can we move forward on that?

**Comment by moulins** [comment]
Having the capability to work with Forge profiles would be a welcome addition to Amidst, but − as you've seen yourself − there's lots of work to be done in the Amidst codebase for this to be possible.

A remote interface seems the way to go for the long-term, but before implementing it, I think the `MinecraftInterface` should be improved, and become the source of truth for the following:
 - the `Biome` list (so that worlds with custom biomes are correctly displayed);
 - the list of compatible biomes for each structure supported by Amidst (so that structure locations are correct on worlds with custom biomes);
 - in the long-term, we could even consider adding support for custom structures/and or dimensions.

I'm not well-versed on how mods handle their custom biomes and structures, so what does Forge need to be able to access this information? Is creating the world sufficient?

**Comment by fraenkelc** [comment]
Depending on the Mod in question Biome (and Structure) data is available either after starting minecraft (for example [Traverse](https://www.curseforge.com/minecraft/mc-mods/traverse-legacy-continued)) or only after creating a world in game ([OTG](https://www.curseforge.com/minecraft/mc-mods/open-terrain-generator), [Climate Control](https://www.curseforge.com/minecraft/mc-mods/climate-control-geographicraft)). The later usually bring their own world type and associated configuration. As such I currently support two modes of operation in the mod: 
* One creates a world with the seed specified in amidst in the minecraft process and gets the data from it. 
* The second connects ignores the values in Amidst(seed, world type) and connects to the running minecraft instance and retrieves data from the live worldserver.

I agree that making the MinecraftInterface the source of truth is an important step.

While having accurate structure information would be nice it is far from essential. Amidst for Forge is currently only offering the biome mapping feature of Amidst. Even with those limitations there are quite a number of users that use it. (I'm not arguing against the implementation; I just think it does not have to be part of the MVP)

**Comment by fraenkelc** [comment]
I've updated my work for forge and minecraft 1.14.4. the current code does not run amidst and minecraft in the same process anymore but transfers the map data using a tcp socket. See https://github.com/toolbox4minecraft/amidst/compare/v4.4-beta1...fraenkelc:forge-mc-1.14

**Comment by moulins** [comment]
@fraenkelc I don't know if you're still working on this project, but there have been some recent changes to the Amidst codebase which may be of interest to you:

- #726 (Version based biomes) removed Amidst's static biome list and introduced a dynamic `BiomeList` class. This biome list is provided as a `VersionFeature` for now, but it should'nt be too difficult to move this to the `MinecraftInterface` itself.
- I've just pushed #750 (Add support for multiple worlds on a single MinecraftInterface), which I'll be merging very soon; this PR greatly change `MinecraftInterface`'s methods, and may require updates to your RPC interface.

If you have any remarks on this, feel free to share them!

**Comment by fraenkelc** [comment]
Thanks for the update. I'm not very active adding new features but I do fixes and ports to new versions now and then. The `BiomeList` extraction definitely makes the forge integration cleaner as I don't have to patch the Biome list by instantiating new biomes. 
https://github.com/fraenkelc/amidst/commit/56e65af3f46ca2e2bb8ac4940ca664f44a9c43e6 moves the `BiomeList` to the `MinecraftInterface`. I moved the whole `VersioneFeature` over as otherwise I'd probably have to break up `DefaultBiomes.DEFAULT_BIOMES` to generate a concrete `BiomeList` for that particular Interface / World.

My current work is WIP - I need to find an acceptable way of fixing up the `BiomeProfileSelection` with an updated `BiomeProfile` that contains colors for all new biomes still.

The interface changes in the branch look manageable. For the remote interface the World interface could use a close() method so that I can also discard the world and resources in the remote minecraft instance once they are no longer used in amidst as garbage collection will not work in this case (across processes).

**Comment by moulins** [comment]
> For the remote interface the World interface could use a close() method so that I can also discard the world and resources in the remote minecraft instance once they are no longer used in amidst

Good catch! I'm not sure how to integrate this in the Amidst architecture however, I'll have to think about it.
